spring-pulsar/src/main/java/org/springframework/pulsar/core/reactive/ReactivePulsarSenderTemplate.java

```diff
@@ -79,22 +91,44 @@ public void setSchema(Schema<T> schema) {
 	private Mono<MessageId> doSend(String topic, T message,
 			MessageSpecBuilderCustomizer<T> messageSpecBuilderCustomizer, MessageRouter messageRouter,
 			ReactiveMessageSenderBuilderCustomizer<T> customizer) {
-		final String topicName = ReactiveMessageSenderUtils.resolveTopicName(topic, this.reactiveMessageSenderFactory);
-		this.logger.trace(() -> String.format("Sending reative msg to '%s' topic", topicName));
+		return doSendMany(topic, Mono.just(message), messageSpecBuilderCustomizer, messageRouter, customizer).single();
+	}
 
-		final ReactiveMessageSender<T> sender = createMessageSender(topic, message, messageRouter, customizer);
+	private Flux<MessageId> doSendMany(String topic, Publisher<T> messages,
+			MessageSpecBuilderCustomizer<T> messageSpecBuilderCustomizer, MessageRouter messageRouter,
+			ReactiveMessageSenderBuilderCustomizer<T> customizer) {
+		final String topicName = ReactiveMessageSenderUtils.resolveTopicName(topic, this.reactiveMessageSenderFactory);
+		this.logger.trace(() -> String.format("Sending reactive msg to '%s' topic", topicName));
+
+		if (this.schema != null) {
```

Comment:
If the template has a schema, we can create the message sender right away and use `ReactiveMessageSender ::sendMessages` to send them as a stream.
Otherwise we need to wait to get a message to create it and we can't share it between messages. So we create one each time and use `ReactiveMessageSender ::sendMessage` to send messages individually.

Reply:
Thanks for the great explanation. I think this would be a great thing to add as a comment.

Reply:
done